vision/dcgan_mnist/dcgan.jl

```diff
@@ -0,0 +1,130 @@
+using Base.Iterators: partition
+using Flux
+using Flux.Optimise: update!
+using Flux: logitbinarycrossentropy
+using Images
+using MLDatasets
+using Statistics
+using Parameters: @with_kw
+using Printf
+
+@with_kw struct HyperParams
+    batch_size::Int = 128
+    latent_dim::Int = 100
+    epochs::Int = 20
+    verbose_freq::Int = 1000
+    output_x::Int = 6
+    output_y::Int = 6
+    lr_dscr::Float64 = 0.0002
+    lr_gen::Float64 = 0.0002
+end
+
+function create_output_image(gen, fixed_noise, hparams)
+    @eval Flux.istraining() = false
+    fake_images = @. cpu(gen(fixed_noise))
+    @eval Flux.istraining() = true
+    image_array = dropdims(reduce(vcat, reduce.(hcat, partition(fake_images, hparams.output_y))); dims=(3, 4))
+    image_array = @. Gray(image_array + 1f0) / 2f0
+    return image_array
+end
+
+# Loss functions
+function discriminator_loss(real_output, fake_output)
+    real_loss = mean(logitbinarycrossentropy.(real_output, 1f0))
+    fake_loss = mean(logitbinarycrossentropy.(fake_output, 0f0))
+    return real_loss + fake_loss
+end
+
+generator_loss(fake_output) = mean(logitbinarycrossentropy.(fake_output, 1f0))
+
+function train_discriminator!(gen, dscr, batch, opt_dscr, hparams)
+    noise = randn(Float32, hparams.latent_dim, hparams.batch_size) |> gpu
```

Comment:
this is not ideal, we should avoid data transfer and use `CuArrays.randn` if training on gpu. I don't know what would be an elegant way to do that, maybe we can revisit in the future

Reply:
we might write something like
```julia
randn!(similar(batch, (hparams.latent_dim, hparams.batch_size)))
```
https://docs.julialang.org/en/v1/stdlib/Random/#Random.randn!, https://github.com/JuliaGPU/CuArrays.jl/blob/da2389f0db46d611a5b625c2ce6bba79aa20163f/src/rand/random.jl#L88-L95

Reply:
nice!